src/blog/0.3/linear-types-borrowing.vmd

```diff
@@ -0,0 +1,673 @@
+
+
+I was surprised to find that the sample program had almost _zero_ memory-safety overhead, except for a line in this ([paraphrased](https://github.com/Verdagon/RegionsBenchmarks/blob/main/cellular-automata/CellularAutomata.vale)) code:
+
+[# E: These examples clear enough? Trying to balance noisiness and explanation, so was sparse on comments.]
+
+```
+func RandInt(rand &Random) i64 {
+  x = rand.num;
+  set x = x + (x / 200096i64);
+  set x = x - (x * 33554432i64);
+  set x = x + (x / 134217728i64);
+  set rand.num = x;
+  return x;
+}
+
+// Returns a 2D array of random bools.
+func MakeBoard(
+    rand &Random,
+    num_rows int,
+    num_cols int)
+[][]bool {
+  rows = [][]bool(num_rows);
+  foreach row_i in 0..num_rows {
+    row = []bool(num_cols);
+    foreach col_i in 0..num_cols {
+      r = RandInt(rand);
+      row.push((r mod 2) == 0);
+    }
+    rows.push(row);
+  }
+  return rows;
+}
+```
+
+
+Specifically, when `RandInt`'s `rand.num` dereferences the `rand` object, it does an implicit [generation check](/blog/generational-references) assertion to ensure the object is still alive.
+
+
+
+[# E: This paragraph is here because I'm trying to convey that generation checks _aren't_ expensive, and not really worth optimizing away much of the time, but we *can* if we want to.]
+
+Normally, I don't believe in chasing zero-cost memory safety. Playing overhead golf can often lead to chasing diminishing returns or prematurely optimizing. After all, a generation check is usually only expensive as a bounds check (in other words, usually negligible).
+
+
+But still, I wondered if it _could_ be done. Could we get rid of that last little bit of overhead?
+
+
+A few weeks later, I discovered the solution.
+
+
+<slice/>
+
+# Moving into and out of functions
+
+<<<<
+Instead of taking a `&Random` reference parameter like this:
+
+```
+func RandInt(rand &Random) i64 {
+  x = rand.num;
+  set x = x + (x / 200096i64);
+  set x = x - (x * 33554432i64);
+  set x = x + (x / 134217728i64);
+  set rand.num = x;
+  return x;
+}
+```
+////
+`RandInt` could take (and return) an owned `Random`, like this:
+
+```
+func RandInt(rand Random) (i64, Random) {
+  x = rand.num;
+  set x = x + (x / 200096i64);
+  set x = x - (x * 33554432i64);
+  set x = x + (x / 134217728i64);
+  set rand.num = x;
+  return (x, rand);
+}
+```
+>>>>
+
+
+Here's what we changed:
+
+ * `rand &Random` became `rand Random`. Instead of taking a reference, the caller moves the actual Random instance itself.
+ * `return x` became `return (x, rand);` which returns a two-element tuple, which also has the Random instance.
+ * The return type `i64` became `(i64, Random)` accordingly.
+
+
+Now, `rand.num` isn't dereferencing a reference, it's accessing a field of an owned value.
+
+Vale's compiler sees this, and doesn't insert a generation check, because it knows the object is alive.
```

Comment:
consider adding "since we own it", feel free to ignore if you think this is self-evident enough